### PR TITLE
[varlen] Fix cuDNN backend bugs found by fuzz testing

### PR DESCRIPTION
**Pull request by drisspg**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* #194834
* __->__ #194818

This commit was authored with an AI assistant.

An adversarial fuzzing campaign against the public varlen_attn cuDNN backend (GB200, cuDNN 9.24.0)
found six reachable defects. Review the kernel-facing fixes in MHA.cpp/attention.cu first, then the Python
selector fix in varlen.py; every fix has a regression test. All new validation is host-side tensor metadata
only: no fix introduces a device-to-host or host-to-device sync.

cuDNN's varlen kernels require 16-byte-aligned base pointers and row strides, but the alignment check only
guarded the KV-cache path. Unaligned q/k/v silently produced wrong results (compute-sanitizer memcheck-clean,
so wrong address arithmetic rather than out-of-bounds access), a 16B-misaligned base pointer raised a sticky
misaligned-address fault, and a padded grad_out crashed backward with CUDNN_BACKEND_API_FAILED err 716. The
forward now enforces has_aligned_varlen_layout for all varlen calls, and backward contiguifies grad_out under
the same predicate. Automatic fallback on this rejection is intentionally not added: the same fuzzing found
Flash varlen can fault on these unaligned layouts, so fallback is not currently a safe recovery path.

The ragged path also trusted more metadata than it checked: a key head dim different from query silently
overread K (proven with a canary allocation), ragged offsets (cum_seqlen * token stride, declared INT32 to
cuDNN) could silently wrap at about 1M packed tokens for H=16/d=128, and empty inputs returned undefined
outputs. These now raise clean errors or return well-defined zero output, -inf LSE, and zero gradients. The
zero-gradient handling covers mixed empty inputs such as nonempty Q with empty KV, where returning from
backward without initializing dQ would otherwise expose allocator contents.

The real cuDNN forward also always allocated a contiguous output, while the fake and Flash forward match the
query's dimension order, so torch.compile with a non-contiguous query failed assert_size_stride. The cuDNN
forward now allocates output in the query's dimension order, restoring parity with Flash and the fake; a
multi-document head-major probe verified this layout in causal/full forward and backward. Finally, head dims
192/256 were incorrectly routed to cuDNN and failed in the frontend, so dimensions above 128 are rejected.

This PR does not claim a performance change. The plot gives backend context on one GB200 by holding the packed
token budget at exactly 16,384 and moving those tokens among six document-length distributions. It uses BF16
causal varlen_attn with H=16, exact triangular matmul FLOPs, and five interleaved timing rounds with
per-case profiler routing checks. PyTorch's registered SDPA FLOP formula is not used because it explicitly
counts the full LxL matrix even for causal attention.

varlen_attn throughput by token-length distribution

FA2 leads when nearly all tokens are in short documents. FA4 takes over as token mass moves into medium/long,
multi-scale, or long-tail distributions; cuDNN generally follows FA4 but loses more on heterogeneous backward
workloads. At d128 uniform-long, median TFLOP/s is 316/685/573 forward and 298/730/660 backward for
FA2/FA4/cuDNN. Raw samples, routing proof, and the benchmark script are in the
[benchmark gist](https://gist.github.com/drisspg/914f2970e21a7661e508184d9fa35426).

```bash
gpu-run auto -- env PYTHONPATH=$PWD /home/drisspg/.venvs/dev/bin/pytest -q test/test_varlen_attention.py -n 8
lintrunner -a

gpu-run auto -- env PYTHONPATH=$PWD /home/drisspg/.venvs/dev/bin/python agent_space/varlen_cudnn_audit/round2/perf_distributions/benchmark.py
```

382 passed / 72 skipped / 8 xfailed on GB200 with cuDNN 9.24.0 (baseline before this change: 372 passed);
the ten new tests cover each fix and fail or expose undefined behavior without it.

cc @liangel-02 @howardzhang-cv

### CODE REVIEW

**Comment by drisspg** [comment]
@claude fable review dis good

File: torch/nn/attention/varlen.py [hunk 1]
```diff
@@ -270,7 +272,16 @@ def _varlen_attn_fake(
     """
     window_size = _normalize_window_size(window_size)
 
-    output = _empty_with_matching_layout(query, (*query.shape[:-1], value.size(-1)))
+    if backend == _CUDNN_ATTENTION_BACKEND:
+        # The cuDNN varlen forward always allocates a contiguous output
+        # (aten/src/ATen/native/cudnn/MHA.cpp), regardless of query layout.
```
Comment (drisspg) [inline torch/nn/attention/varlen.py]: we should fix this

File: aten/src/ATen/native/cudnn/MHA.cpp [hunk 2]
```diff
@@ -581,6 +609,43 @@ std::vector<int64_t> thd_to_bhsd_strides(const Tensor& tensor) {
   return {INT_MAX, tensor.stride(1), tensor.stride(0), tensor.stride(2)};
 }
 
+// Allocate a dense output in the query's dimension order, mirroring
+// _empty_with_matching_layout in torch/nn/attention/_utils.py (used by the
+// varlen custom-op fake) and the Flash varlen forward's empty_like(q).
+Tensor empty_with_matching_layout(const Tensor& tensor, IntArrayRef shape) {
```
Comment (eqy) [inline aten/src/ATen/native/cudnn/MHA.cpp]: Is this reconcilable with https://github.com/pytorch/pytorch/blob/60598ed3c8773875c0923101d54f206303b2f59f/aten/src/ATen/native/transformers/sdp_utils.h#L7 ?

File: aten/src/ATen/native/cudnn/MHA.cpp [hunk 3]
```diff
@@ -581,6 +609,43 @@ std::vector<int64_t> thd_to_bhsd_strides(const Tensor& tensor) {
...
+  for (const auto dim : order) {
+    strides[dim] = stride;
+    stride *= shape[dim];
+  }
+  return at::empty_strided(shape, strides, tensor.options());
+}
+
+// Ragged offsets are declared int32 to cuDNN, so the largest offset
+// (packed extent * token stride) must not wrap.
+void check_ragged_offset_capacity(const Tensor& tensor, const char* name) {
```
Comment (eqy) [inline aten/src/ATen/native/cudnn/MHA.cpp:613]: does it make sense to make this inline

File: test/test_varlen_attention.py [hunk 4]
```diff
@@ -974,6 +974,196 @@ def test_cudnn_causal_varlen_requires_shared_cu_seq(self, device):
...
+
+        q, k, v = (make(i == tensor_idx) for i in range(3))
+        cu_seq = torch.tensor([0, seq_len], device=device, dtype=torch.int32)
+        with sdpa_kernel(SDPBackend.CUDNN_ATTENTION):
+            with self.assertRaisesRegex(RuntimeError, "16-byte-aligned"):
+                varlen_attn(q, k, v, cu_seq, cu_seq, seq_len, seq_len)
+
+    @skipIfRocm
+    @unittest.skipIf(
+        not PLATFORM_SUPPORTS_FLASH_ATTENTION, "Flash Attention not supported"
```
Comment (eqy) [inline test/test_varlen_attention.py:1007]: Is the the correct decorator or does this need the cudnn platform supports cudnn decorator

File: test/test_varlen_attention.py [hunk 5]
```diff
@@ -974,6 +974,196 @@ def test_cudnn_causal_varlen_requires_shared_cu_seq(self, device):
         self.assertEqual(cudnn_forward.call_count, 0)
         self.assertEqual(cudnn_backward.call_count, 0)
 
+    @skipIfRocm
+    @parametrize("tensor_idx", [0, 1, 2])
```
Comment (eqy) [inline test/test_varlen_attention.py:978]: does this need the platform supports cudnn decorator etc

**Comment by eqy** [comment]
@claude fable do the new human review comments make sense?

File: aten/src/ATen/native/cudnn/MHA.cpp (hunk 3 above)
Comment (drisspg) [reply aten/src/ATen/native/cudnn/MHA.cpp:613]: ehh its in anon namespace and I dont subscribe to `big inline`

File: test/test_varlen_attention.py (hunk 4 above)
Comment (drisspg) [reply test/test_varlen_attention.py:1007]: varlen has stricter cudnn reqs so we just skip on first line

File: aten/src/ATen/native/cudnn/MHA.cpp (hunk 3 above)
Comment (eqy) [reply aten/src/ATen/native/cudnn/MHA.cpp:613]: "i like big callstacks and i cannot lie"

File: aten/src/ATen/native/cudnn/MHA.cpp [hunk 6]
```diff
@@ -1976,29 +2031,37 @@ void run_cudnn_SDP_bprop_nestedtensor(
...
   auto softmaxstats_ = softmaxstats.unsqueeze(-1).transpose(0, 1);
 
-  // Alignment is not part of the cache key, and cuDNN requires a unit
-  // embedding stride. Preserve all other dO strides when those hold.
+  // Alignment is not part of the cache key, and cuDNN requires 16-byte
+  // pointer/stride alignment. Preserve dO strides when those hold.
   Tensor dO_ = dO;
-  if (dO.stride(-1) != 1 ||
-      reinterpret_cast<uintptr_t>(dO.const_data_ptr()) % 16 != 0) {
+  if (!has_aligned_varlen_layout(dO)) {
```
Comment (eqy) [inline aten/src/ATen/native/cudnn/MHA.cpp:2048]: interesting that this is the only ptr with this requirement

File: aten/src/ATen/native/cudnn/MHA.cpp [hunk 7]
```diff
@@ -1976,29 +2031,37 @@ void run_cudnn_SDP_bprop_nestedtensor(
...
     dO_ = dO.clone(at::MemoryFormat::Contiguous);
   }
   TORCH_INTERNAL_ASSERT(
-      dO_.stride(-1) == 1 &&
-          reinterpret_cast<uintptr_t>(dO_.const_data_ptr()) % 16 == 0,
+      has_aligned_varlen_layout(dO_),
       "cuDNN SDPA expected an aligned grad_output with unit embedding stride");
```
Comment (eqy) [inline aten/src/ATen/native/cudnn/MHA.cpp]: is this message still valid?

**Review by eqy — APPROVED** [review]

File: aten/src/ATen/native/cudnn/MHA.cpp (hunk 6 above)
Comment (drisspg) [reply aten/src/ATen/native/cudnn/MHA.cpp:2048]: only one we dont control / is not validated before entry
